File: apps/synqra-mvp/lib/council/groq-client.ts
import { councilConfig } from "./config";
import type { GroqResult, GroqUsage } from "./types";

const GROQ_ENDPOINT = "https://api.groq.com/openai/v1/chat/completions";
const GROQ_HARD_MAX_TOKENS = councilConfig.maxTokens;
const GROQ_DEFAULT_MAX_TOKENS = Math.min(councilConfig.maxTokens, 280);
const GROQ_MAX_PROMPT_CHARS = 1800;

type GroqMessage = { role: "system" | "user"; content: string };

type GroqResponse = {
  choices?: Array<{ message?: { content?: string | null } }>;
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
  };
  model?: string;
};

type GroqUsageSnapshot = {
  calls: number;
  success: number;
  inputTokens: number;
  outputTokens: number;
};

const groqUsageSnapshot: GroqUsageSnapshot = {
  calls: 0,
  success: 0,
  inputTokens: 0,
  outputTokens: 0,
};

function buildUsage(data?: GroqResponse["usage"]): GroqUsage {
  return {
    inputTokens: data?.prompt_tokens ?? 0,
    outputTokens: data?.completion_tokens ?? 0,
  };
}

function getGroqApiKey(): string | null {
  const value = process.env.GROQ_API_KEY?.trim();
  return value ? value : null;
}

function sanitizeGroqError(text: string): string {
  return text.replace(/\s+/g, " ").trim().slice(0, 240);
}

function updateGroqUsage(usage: GroqUsage): void {
  groqUsageSnapshot.success += 1;
  groqUsageSnapshot.inputTokens += usage.inputTokens;
  groqUsageSnapshot.outputTokens += usage.outputTokens;
}

export function hasGroqApiKey(): boolean {
  return Boolean(getGroqApiKey());
}

export function getGroqUsageSnapshot(): GroqUsageSnapshot {
  return { ...groqUsageSnapshot };
}

export async function requestGroq(options: {
  prompt: string;
  systemPrompt?: string;
  temperature?: number;
  maxTokens?: number;
  model?: string;
  signal?: AbortSignal;
}): Promise<GroqResult> {
  groqUsageSnapshot.calls += 1;

  const apiKey = getGroqApiKey();
  if (!apiKey) {
    return {
      success: false,
      error: { code: "groq_unavailable", message: "GROQ_API_KEY is missing." },
    };
  }

  const prompt = options.prompt.trim();
  if (!prompt) {
    return {
      success: false,
      error: { code: "groq_invalid_prompt", message: "Prompt is required." },
    };
  }
  if (prompt.length > GROQ_MAX_PROMPT_CHARS) {
    return {
      success: false,
      error: {
        code: "groq_prompt_too_large",
        message: `Prompt exceeds ${GROQ_MAX_PROMPT_CHARS} characters.`,
      },
    };
  }

  const messages: GroqMessage[] = [];
  if (options.systemPrompt) {
    messages.push({ role: "system", content: options.systemPrompt });
  }
  messages.push({ role: "user", content: prompt });

  try {
    const maxTokens = Math.min(options.maxTokens ?? GROQ_DEFAULT_MAX_TOKENS, GROQ_HARD_MAX_TOKENS);
    const response = await fetch(GROQ_ENDPOINT, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model: options.model?.trim() || councilConfig.model,
        messages,
        temperature: options.temperature ?? councilConfig.temperature,
        max_tokens: maxTokens,
      }),
      signal: options.signal,
    });

    if (!response.ok) {
      const text = sanitizeGroqError(await response.text());
      return {
        success: false,
        error: {
          code: "groq_failed",
          message: `Groq request failed (${response.status}): ${text || "unknown error"}`,
        },
      };
    }

    const data = (await response.json()) as GroqResponse;
    const content = data.choices?.[0]?.message?.content?.trim();

    if (!content) {
      return {
        success: false,
        error: { code: "groq_failed", message: "Groq returned empty response." },
      };
    }

    const usage = buildUsage(data.usage);
    updateGroqUsage(usage);

    return {
      success: true,
      content,
      model: data.model ?? options.model?.trim() ?? councilConfig.model,
      usage,
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : "Unknown Groq error";
    return {
      success: false,
      error: { code: "groq_failed", message },
    };
  }
}

File: apps/synqra-mvp/services/discovery_worker/groq_client.py
import os
from typing import Any
from urllib.request import Request, urlopen

GROQ_CHAT_URL = "https://api.groq.com/openai/v1/chat/completions"
DEFAULT_GROQ_MODEL = "llama-3.3-70b-versatile"


class GroqEnrichmentClient:
    def __init__(self, api_key: str | None = None, model: str | None = None, timeout_seconds: float = 20.0) -> None:
        self._api_key = (api_key or os.getenv("GROQ_API_KEY", "")).strip()
        self._model = (model or os.getenv("GROQ_MODEL", DEFAULT_GROQ_MODEL)).strip() or DEFAULT_GROQ_MODEL
        self._timeout_seconds = timeout_seconds

    @property
    def model_name(self) -> str:
        return self._model

    def is_configured(self) -> bool:
        return bool(self._api_key)

    def generate_enrichment(self, prompt: str) -> dict[str, Any]:
        if not self._api_key:
            raise RuntimeError("GROQ_API_KEY is not configured")

        payload = {
            "model": self._model,
            "messages": [{"role": "user", "content": prompt}],
            "temperature": 0.2,
        }
        import json

        request = Request(
            GROQ_CHAT_URL,
            data=json.dumps(payload).encode("utf-8"),
            method="POST",
            headers={"Authorization": f"Bearer {self._api_key}", "Content-Type": "application/json"},
        )
        with urlopen(request, timeout=self._timeout_seconds) as response:
            body = response.read().decode("utf-8")
        response = json.loads(body)
        return _extract_json_from_response(response)


def _extract_json_from_response(response: dict[str, Any]) -> dict[str, Any]:
    choices = response.get("choices")
    if not isinstance(choices, list) or not choices:
        raise RuntimeError("Groq response missing choices")
    first = choices[0]
    message = first.get("message") if isinstance(first, dict) else None
    content = message.get("content") if isinstance(message, dict) else None
    if not isinstance(content, str):
        raise RuntimeError("Groq response missing message content")

    import json

    return json.loads(content)
